larray/core/metadata.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf8 -*-
+from __future__ import absolute_import, division, print_function
+
+from collections import OrderedDict
+from larray.util.misc import PY2
+
+
+__all__ = ['Metadata']
+
+
+class BaseMetadata(OrderedDict):
+    """
+    Base class for AttributeDict subclass.
+
+    Examples
+    --------
+    >>> import os.path
+    >>> from datetime import datetime
+    >>> from larray.core.array import ndtest
+    >>> tmpdir = getfixture('tmpdir')
+    >>> attrs = Metadata([('string', 'a string'), ('number', 5), ('float', 5.6), ('time', datetime(2017, 2, 10))])
+
+    Save/Load using HDF format
+
+    >>> from larray.util.misc import LHDFStore
+    >>> fname = os.path.join(tmpdir.strpath, 'metadata.h5')
+    >>> key = 'arr'
+    >>> with LHDFStore(fname) as store:
+    ...     # save
+    ...     ndtest(3).to_hdf(store, key)
+    ...     attrs.to_hdf(store, key)
+    ...     # load
+    ...     attrs2 = BaseMetadata.from_hdf(store, key)
+    >>> attrs2
+    string: a string
+    number: 5
+    float: 5.6
+    time: 2017-02-10 00:00:00
+    """
+    def __init__(self, *args, **kwargs):
+        super(BaseMetadata, self).__init__(*args, **kwargs)
+
+    def __repr__(self):
+        return '\n'.join(['{}: {}'.format(k, v) for k, v in self.items()])
+
+    # ---------- IO methods ----------
+    def to_hdf(self, hdfstore, key):
+        if len(self):
+            hdfstore.get_storer(key).attrs.metadata = self
+
+    @classmethod
+    def from_hdf(cls, hdfstore, key):
+        if 'metadata' in hdfstore.get_storer(key).attrs:
+            return hdfstore.get_storer(key).attrs.metadata
+
+
+if PY2:
```

Comment:
having different functionality (and docstring) on Python2 bothers me. Why don't you use a instance of ordered dict as an attribute on Python 2? Something like:

```python
class Metadata(object):
    def __init__(self, *args, **kwargs):
        object.__setattr__(self, '__odict', OrderedDict(*args, **kwargs))

    def __getattr__(self, key):
        od = object.__getattribute__(self, '__odict')
        if hasattr(od, key):
            return getattr(od, key)
        else:
            try:
                return od[key]
            except KeyError:
                raise AttributeError(key)

    def __setattr__(self, key, value):
        od = object.__getattribute__(self, '__odict')
        od[key] = value

    def __delattr__(self, key):
        od = object.__getattribute__(self, '__odict')
        del od[key]

    def method_factory(name):
        fullname = '__%s__' % name
        odict_method = getattr(OrderedDict, fullname)
        def method(self, *args, **kwargs):
            od = object.__getattribute__(self, '__odict')
            return odict_method(od, *args, **kwargs)
        return method

    __getitem__ = method_factory('getitem')
    __setitem__ = method_factory('setitem')
    __delitem__ = method_factory('delitem')
    __contains__ = method_factory('contains')

    __dir__ = method_factory('dir')

    __eq__ = method_factory('eq')
    __ne__ = method_factory('ne')
    __ge__ = method_factory('ge')
    __gt__ = method_factory('gt')
    __le__ = method_factory('le')
    __lt__ = method_factory('lt')

    __iter__ = method_factory('iter')
    __len__ = method_factory('len')

    __reduce__ = method_factory('reduce')
    __reduce_ex__ = method_factory('reduce_ex')

    __reversed__ = method_factory('reversed')

    __sizeof__ = method_factory('sizeof')

    def __repr__(self):
        return '\n'.join(['{}: {}'.format(k, v) for k, v in self.items()])
```
TODO: 
* the binary op methods do not work (because it translates `Metadata().__eq__(Metadata())` to `Metadata().__odict.__eq__(Metadata())` and OrderedDict does not understand Metadata objects (this is simple to fix though).
* `__dir__` should be overridden to return keys too (on Python3 too)
* sizeof, reduce and reduce_ex are probably broken. This is not very important though and should not be too hard to fix.